### PR TITLE
Add hierarchy structure for project nodes in task view

### PR DESCRIPTION
**Pull request by Camork**

### CODE REVIEW

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java [hunk 1]
```diff
@@ -71,9 +71,20 @@ public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
         // flatten the tree of Gradle projects to a list, similar
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: This comment is no longer valid, please adjust it reflecting to the `if` statement below.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java [hunk 2]
```diff
@@ -71,9 +71,20 @@ public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
...
         // to how Eclipse projects look in the Eclipse Project explorer
         List<BaseProjectNode> allProjectNodes = Lists.newArrayList();
-        for (OmniEclipseProject project : projects) {
-            if (project.getParent() == null) {
-                collectProjectNodesRecursively(project, null, allProjectNodes);
+        
+        if(this.taskView.getState().isGroupProjects()) {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: I think we could use a better naming here. It should express that the project hierarchy is (or is not) flattened in the Tasks view. How about `TaskViewState.isProjectHierarchyFlattened()`?

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/GroupProjectsAction.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import org.eclipse.buildship.ui.PluginImage;
+import org.eclipse.buildship.ui.PluginImages;
+import org.eclipse.jface.action.Action;
+
+/**
+ * An action on the {@link TaskView} that toggles project grouping.
+ *
+ * @author Charles Wu <charles.wu@liferay.com> 
+ */
+public final class GroupProjectsAction extends Action {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/GroupProjectsAction.java]: Rename to `ShowFlattenedProjectHierarchy` and verify that the implementation matches the naming.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/GroupProjectsAction.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        this.taskView = Preconditions.checkNotNull(taskView);
+
+        setText(TaskViewMessages.Action_GroupProjectsByHierarchy_Text);
+        setImageDescriptor(PluginImages.PROJECT_GROUP.withState(PluginImage.ImageState.ENABLED).getImageDescriptor());
+        setChecked(this.taskView.getState().isGroupProjects());
+    }
+
+    @Override
+    public void run() {
+        this.taskView.getState().setGroupProjects(isChecked());
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/GroupProjectsAction.java]: `setProjectHierarchyFlattened()`

File: org.eclipse.buildship.ui/src/main/resources/org/eclipse/buildship/ui/view/task/TaskViewMessages.properties [hunk 5]
```diff
@@ -38,4 +38,5 @@ Action_FilterTaskSelectors_Text=Show Task Selectors
 Action_FilterPrivateTasks_Text=Show All Tasks
 Action_SortByType_Text=Order Tasks by Type
 Action_SortByVisibility_Text=Order Tasks by Visibility
+Action_GroupProjectsByHierarchy_Text=Group Projects By Hierarchy
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/resources/org/eclipse/buildship/ui/view/task/TaskViewMessages.properties]: `Action_ShowFlattenProjectHiearchy=Flatten Project Hierarchy`

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java [hunk 6]
```diff
@@ -131,11 +142,34 @@ public boolean hasChildren(Object element) {
     }
 
     private Object[] childrenOf(ProjectNode projectNode) {
+        boolean isGroupProjects = this.taskView.getState().isGroupProjects();
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: Please clean this up a bit. This method should look something like this: 
    
    if (this.taskView.getState().isGroupProjects()) {
        result.addAll(projectNodesFor(projectNode));
    }
    if (this.taskView.getState().isGroupTasks()) {
            result.addAll(groupNodesFor(projectNode));
    } else {
        result.addAll(taskNodesFor(projectNode));
    }

The `projectNodesFor()` method should look very similar to `taskNodesFor()`.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java [hunk 7]
```diff
@@ -131,11 +142,34 @@ public boolean hasChildren(Object element) {
...
+        return projectNodesFor(projectNode, isGroupProjects).toArray();
+    }
+
+    private Set<Object> projectNodesFor(ProjectNode projectNode, boolean isGroupProjects) {
+        Set<Object> result = Sets.newHashSet();
+
+        if (isGroupProjects) {
+            OmniEclipseProject eclipseProject = projectNode.getEclipseProject();
+            OmniGradleProject gradleProject;
+            Optional<IProject> workspaceProject;
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: Why not initialize the variables inside the `for` loop? Anyway, they might disappear in the refactoring.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java [hunk 8]
```diff
@@ -54,6 +56,7 @@ public void load() {
         this.linkToSelection = prefs.getBoolean(PREF_LINK_TO_SELECTION, false);
         this.headerNameColumnWidth = prefs.getInt(PREF_HEADER_NAME_COLUMN_WIDTH, 200);
         this.headerDescriptionColumnWidth = prefs.getInt(PREF_HEADER_DESCRIPTION_COLUMN_WIDTH, 400);
+        this.groupProjects = prefs.getBoolean(PREF_GROUP_PROJECTS_BY_HIERARCHY, true);
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java]: Making the hierarchical project representation default breaks the test coverage. I ran the build on our CI, you should be able to check the results [here](https://builds.gradle.org/project.html?projectId=Tooling_Buildship&tab=projectOverview). (credentials: guest/guest). If not, please let me know.

Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java]: Please add test coverage to `TaskViewContentTest` for the new hierarchy.

The Oomph buildship setup should give you a working IDE setup. Just a couple of notes:
- sometimes you have to activate the `tooling-e47.target` target definition file to compile everything.
- the UI tests are launched on the UI thread by default which causes the Buildship UI tests to hang. You have to disable it in the run configurations to make them work.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java (hunk 7 above)
Comment (Camork) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: I had a consideration in performance perspective. Actually, it's not good for readable.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java (hunk 2 above)
Comment (Camork) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewContentProvider.java]: Yeah

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java (hunk 8 above)
Comment (Camork) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java]: Thanks for this tips, I will add this tests in recent days

Comment (Camork) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java]: I am done.

Comment (donat) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskViewState.java]: Awesome! I'll soon check the latest changes.

**Comment by donat** [comment]
@Camork Thank you for the contribution!

**Comment by Camork** [comment]
Hope to have contribution again :）

**Comment by regrog** [comment]
I think that maybe the order should not mix subprojects with groups.
In another case groups are splitted even more.
I also see that projects are not ordered by name.
image

**Comment by Camork** [comment]
The order problem had fixed in branch 3.0 https://github.com/eclipse/buildship/pull/717, but not merged into 2.2(master).

**Pull request merged**